File: 2316-count-hills-and-valleys-in-an-array/2316-count-hills-and-valleys-in-an-array.py
class Solution:
    def countHillValley(self, nums: List[int]) -> int:
        hills_valleys_count = 0
        n = len(nums)

        # Edge Case;
        if n <= 2:
            return 0

        for idx in range(1, n - 1):
            curr_num = nums[idx]

            prev_idx = idx - 1
            # skip the same valley or hill;
            if curr_num == nums[prev_idx]:
                continue

            while prev_idx >= 0 and nums[prev_idx] == curr_num:
                prev_idx -= 1
            
            next_idx = idx + 1
            while next_idx < n and nums[next_idx] == curr_num:
                next_idx += 1
            
            if prev_idx < 0 or next_idx == n:
                continue

            # print(curr_num)
            # else did we find a hill or valley;
            if nums[prev_idx] < curr_num > nums[next_idx]:
                # case for hill
                hills_valleys_count += 1
            elif nums[prev_idx] > curr_num < nums[next_idx]:
                # case for hill
                hills_valleys_count += 1
        
        return hills_valleys_count
                


        

File: 2316-count-hills-and-valleys-in-an-array/2316-count-hills-and-valleys-in-an-array.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var countHillValley = function(nums) {
    // Observation:
    // Both the non-equal neighbors should be smaller to be a hill
    // Both the non-equal neighbors should be larger to be a valley
    // Same hill or valley if value at adjacent indexes i & j are same i.e nums[i] === nums[j] 

    let totalHillValleyCount = 0;
    const numsArrLen = nums.length;

    const firstElIdx = 1;
    const lastElIdx = numsArrLen - 1;

    for(let idxI = firstElIdx; idxI < lastElIdx; idxI++){
        let nonDuplicateElIdx = idxI + 1;

        while(nonDuplicateElIdx < numsArrLen && nums[idxI] === nums[nonDuplicateElIdx]){
            nonDuplicateElIdx++;
        }

        if(nonDuplicateElIdx === numsArrLen){
            return totalHillValleyCount;
        }

        const prevNonDuplicateEl = nums[idxI - 1];
        const currEl = nums[idxI];
        const nextNonDuplicateEl = nums[nonDuplicateElIdx];

        // console.log(prevNonDuplicateEl, currEl, nextNonDuplicateEl)

        const isHillNeighborsSmaller = (currEl > prevNonDuplicateEl) && (currEl > nextNonDuplicateEl);
        const isValleyNeighborsLarger = (currEl < prevNonDuplicateEl) && (currEl < nextNonDuplicateEl);

        if(isHillNeighborsSmaller || isValleyNeighborsLarger){
            totalHillValleyCount++;
        }

        idxI = nonDuplicateElIdx - 1;
    }

    return totalHillValleyCount;
};